### PR TITLE
Positive and negative agreement (likert) questions

### LINKED ISSUE #1963: Inverted Likert Scale
We currently have the following likert scale, ranging from "Strongly agree" (positive) to "Strongly disagree" (negative):
image

The connotation is shown by color. In the results computation, grades reach from 1.0 ("Strongly agree") to 5.0 ("Strongly disagree").

For some questions, an inverted scale would be a better fit (e.g., "I felt overwhelmed by the workload"). Currently, these questions have to be rephrased ("I did not feel overwhelmed ...") which is not optimal.

An inverted likert scale should be introduced, using the range from "Strongly disagree" (green, 1.0) to "Strongly agree" (red, 5.0).
The current 'LIKERT' scale should be renamed to "POSITIVE_LIKERT" ("Positive agreement") and the new scale will be "NEGATIVE_LIKERT" ("Negative agreement").

### PR DESCRIPTION
**Pull request by FSadrieh**
fixes #1963

### CODE REVIEW

File: evap/evaluation/models.py [hunk 1]
```diff
@@ -1081,7 +1081,8 @@ class Question(models.Model):
     """A question including a type."""
 
     TEXT = 0
-    LIKERT = 1
+    POSITIVE_LIKERT = 1
+    NEGATIVE_LIKERT = 69
```
Comment (niklasmohrin) [inline evap/evaluation/models.py:1085]: No

File: evap/evaluation/models.py [hunk 2]
```diff
@@ -1094,7 +1095,8 @@ class Question(models.Model):
     HEADING = 5
     QUESTION_TYPES = (
         (_("Text"), ((TEXT, _("Text question")),)),
-        (_("Unipolar Likert"), ((LIKERT, _("Agreement question")),)),
+        (_("Unipolar Likert"), ((POSITIVE_LIKERT, _("Agreement question")),)),
+        (_("Unipolar Likert"), ((NEGATIVE_LIKERT, _("Negative agreement question")),)),
```
Comment (niklasmohrin) [inline evap/evaluation/models.py:1099]: "Disagreement question"? @janno42 @He3lixxx 

Furthermore, "unipolar likert" now appears twice, which seems wrong

File: evap/results/tests/test_tools.py [hunk 3]
```diff
@@ -234,7 +240,9 @@ def test_average_grade(self):
         contributor2_average = 4
         contributors_average = ((4 * contributor1_average) + (2 * contributor2_average)) / (4 + 2)  # 2.9333333
 
-        general_non_grade_average = ((5 * 5) + (3 * 3) + (2 * 5) + (4 * 7 / 3)) / (5 + 3 + 2 + 4)  # 3.80952380
+        general_non_grade_average = ((5 * 5) + (3 * 3) + (2 * 4) + (2 * 5) + (4 * 7 / 3)) / (
+            5 + 3 + 4 + 2 + 4
+        )  # 3.80952380
```
Comment (niklasmohrin) [inline evap/results/tests/test_tools.py:245]: Since the calculation changed, I suppose the result changes too?

**Review by niklasmohrin — CHANGES_REQUESTED** [review]

File: evap/evaluation/models.py (hunk 2 above)
Comment (He3lixxx) [reply evap/evaluation/models.py:1099]: Quote from the issue:
> The current 'LIKERT' scale should be renamed to "POSITIVE_LIKERT" ("Positive agreement") and the new scale will be "NEGATIVE_LIKERT" ("Negative agreement").

I agree with that (positively).

File: evap/evaluation/models.py [hunk 4]
```diff
@@ -1094,7 +1095,8 @@ class Question(models.Model):
     HEADING = 5
     QUESTION_TYPES = (
         (_("Text"), ((TEXT, _("Text question")),)),
-        (_("Unipolar Likert"), ((LIKERT, _("Agreement question")),)),
+        (_("Positive Likert"), ((POSITIVE_LIKERT, _("Positive agreement")),)),
+        (_("Negative Likert"), ((NEGATIVE_LIKERT, _("Negative agreement")),)),
```
Comment (janno42) [inline evap/evaluation/models.py:1099]:
```suggestion
        (
            _("Unipolar Likert"),
            (
                (POSITIVE_LIKERT, _("Positive agreement")),
                (NEGATIVE_LIKERT, _("Negative agreement")),
            ),
        ),
```

**Review by janno42 — CHANGES_REQUESTED** [review]

**Comment by janno42** [comment]
I added a commit with a warning icon for the inverted color scales

File: evap/evaluation/models.py [hunk 5]
```diff
@@ -1105,7 +1106,14 @@ class Question(models.Model):
 
     QUESTION_TYPES = (
         (_("Text"), ((QuestionType.TEXT, _("Text question")),)),
-        (_("Unipolar Likert"), ((QuestionType.LIKERT, _("Agreement question")),)),
+        (
+            _("Unipolar Likert"),
+            (
+                (QuestionType.POSITIVE_LIKERT, _("Positive agreement")),
+                (QuestionType.NEGATIVE_LIKERT, _("Negative agreement")),
+            ),
```
Comment (He3lixxx) [inline evap/evaluation/models.py:1114]: All other questions have the "question" suffix in their description / user-facing name ("Text question", "Positive yes-no question", ...). This distinguishes them from headings. For consistency, I thus think these should be "Positive agreement question" and "Negative agreement question", or is there any reason not to do that?

File: evap/evaluation/tests/test_views.py [hunk 6]
```diff
@@ -183,3 +183,36 @@ def test_edit_display_name(self):
 
         page = self.app.get(self.url, user=self.responsible)
         self.assertContains(page, "testdisplayname")
+
+        form = page.forms["profile-form"]
+        form["first_name_chosen"] = "testdisplayname2"
+        form.submit(name="operation", value="illegal", status=400)
+        self.assertFalse(UserProfile.objects.filter(first_name_chosen="testdisplayname2").exists())
+
+
```
Comment (He3lixxx) [inline evap/evaluation/tests/test_views.py:192]: Changes from another domain unrelated to this PR -- I think something went wrong during merging / rebasing here?

File: evap/evaluation/tests/test_views.py [hunk 7]
```diff
@@ -183,3 +183,36 @@ def test_edit_display_name(self):
...
+        )
+        cls.contribution = baker.make(
+            Contribution,
+            evaluation=cls.evaluation,
+            questionnaires=[cls.question.questionnaire],
+            contributor=cls.contributor,
+        )
+        cls.url = f"/student/vote/{cls.evaluation.pk}"
+
+    def test_likert_questions(self):
```
Comment (He3lixxx) [inline evap/evaluation/tests/test_views.py:216]: maybe `test_answer_ordering`?

File: evap/evaluation/tests/test_views.py [hunk 8]
```diff
@@ -183,3 +183,36 @@ def test_edit_display_name(self):
...
         page = self.app.get(self.url, user=self.responsible)
         self.assertContains(page, "testdisplayname")
+
+        form = page.forms["profile-form"]
+        form["first_name_chosen"] = "testdisplayname2"
+        form.submit(name="operation", value="illegal", status=400)
+        self.assertFalse(UserProfile.objects.filter(first_name_chosen="testdisplayname2").exists())
+
+
+class TestNegativeLikertQuestions(WebTest):
```
Comment (He3lixxx) [inline evap/evaluation/tests/test_views.py:193]: Maybe we should also check for the existence of the warning icon with the tooltip?

**Review by He3lixxx — APPROVED** [review]

**Review by janno42 — CHANGES_REQUESTED** [review]
✔️ Meets requirements
✔️ UI functionality checked

Only some very minor things left from my side

File: evap/evaluation/tests/test_views.py [hunk 9]
```diff
@@ -183,3 +183,35 @@ def test_edit_display_name(self):
...
+            text_en="Negative Likert Question",
+            text_de="Negative Likert Frage",
+        )
+        cls.contribution = baker.make(
+            Contribution,
+            evaluation=cls.evaluation,
+            questionnaires=[cls.question.questionnaire],
+            contributor=cls.contributor,
+        )
+        cls.url = f"/student/vote/{cls.evaluation.pk}"
```
Comment (niklasmohrin) [inline evap/evaluation/tests/test_views.py:212]: Let's use `reverse` instead

File: evap/evaluation/tests/test_views.py [hunk 10]
```diff
@@ -183,3 +183,35 @@ def test_edit_display_name(self):
 
         page = self.app.get(self.url, user=self.responsible)
         self.assertContains(page, "testdisplayname")
+
+
+class TestNegativeLikertQuestions(WebTest):
+    @classmethod
+    def setUpTestData(cls):
+        cls.voting_user = baker.make(UserProfile, email="voting_user1@institution.example.com")
+        cls.contributor = baker.make(UserProfile, email="contributor1@institution.example.com")
```
Comment (niklasmohrin) [inline evap/evaluation/tests/test_views.py:192]: Do you really need a contributor? Shouldn't a question in the general questionnaire be just fine?

**Review by niklasmohrin — COMMENTED** [review]
Code looks mostly good, haven't tested in browser yet

**Review by niklasmohrin — APPROVED** [review]

**Comment by niklasmohrin** [comment]
An example of the new question type can be found in "Beautiful Data" at the end of "Prof. Dr. responsible"

**Review by janno42 — APPROVED** [review]

**Comment by janno42** [comment]
Great teamwork :)
Thanks, @FSadrieh!

**Pull request merged**